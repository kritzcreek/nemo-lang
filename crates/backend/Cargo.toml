[package]
name = "backend"
version = "0.1.0"
edition = "2021"
license = "MIT"

[dependencies]
text-size = "1.1.1"
wasm-encoder = "0.209"
frontend = { path = "../frontend" }

[dev-dependencies]
insta = { version = "1.34.0", features = ["glob"] }
insta-cmd = "0.6.0"
