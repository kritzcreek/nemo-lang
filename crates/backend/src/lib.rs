pub mod codegen;
mod wasm_builder;

use codegen::codegen;
use frontend::run_frontend;

pub fn compile_program(source: &str) -> Result<Vec<u8>, String> {
    let check_result = run_frontend(source);

    if !check_result.errors.is_empty() {
        for err in &check_result.errors {
            eprintln!(
                "{}",
                err.display(source, &check_result.names.name_map, true)
            );
        }
        return Err(format!(
            "Compiling failed with {} errors",
            check_result.errors.len()
        ));
    }

    let (wasm, _) = codegen(
        check_result.ir.expect("No IR despite no check errors"),
        check_result.names,
    );
    Ok(wasm)
}
