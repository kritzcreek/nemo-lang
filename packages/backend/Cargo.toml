[package]
name = "backend"
version = "0.1.0"
edition = "2021"
license = "MIT"

[dependencies]
text-size = "1.1.1"
wasm-encoder = "0.205"
