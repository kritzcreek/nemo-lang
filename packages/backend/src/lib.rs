pub mod codegen;
pub mod ir;
mod names;
mod wasm_builder;
