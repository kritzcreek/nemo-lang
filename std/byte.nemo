module byte
exports (eq, concat)

fn eq(a: bytes, b: bytes) -> bool {
  let len = bytes_len(a);
  if len != bytes_len(b) {
    return false;
  } else {};
  let i = 0;
  while i < len {
    if bytes_get(a, i) != bytes_get(b, i) {
      return false;
    } else {};
    set i = i + 1;
  };
  true
}

fn concat(b1: bytes, b2: bytes) -> bytes {
  let len1 = bytes_len(b1);
  let len2 = bytes_len(b2);
  let len = len1 + len2;
  let b = bytes_new(0, len);
  bytes_copy(b, 0, b1, 0, len1);
  bytes_copy(b, len1, b2, 0, len2);
  b
}
