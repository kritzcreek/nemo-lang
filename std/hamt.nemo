module hamt
exports (Hamt, new, get, add, remove, values, print)

use option
use vec
use string

fn print_hash(hash : I32) -> string::Str {
  let chars : vec::Vec[I32] = vec::new(0);
  let i = 0;
  while i < 32 {
    if i != 0 && i32_rem_u(i, 5) == 0 {
      vec::push(chars, 95)
    } else {};
    if i32_and(hash, 1) == 0 { vec::push(chars, 48) } else { vec::push(chars, 49) };
    set hash = i32_shr_u(hash, 1);
    set i = i + 1;
  };

  let char_count = chars.len;
  let bytes = bytes_new(0, char_count);
  let j = 0;
  while j < char_count {
    bytes_set(bytes, char_count - j - 1, chars.buf[j]);
    set j = j + 1;
  };
  string::unsafe_new(bytes)
}

fn print_bitmap(bitmap : I32) -> string::Str {
  let chars : vec::Vec[I32] = vec::new(0);
  let i = 0;
  while i < 32 {
    if i32_and(bitmap, 1) == 0 { vec::push(chars, 48) } else { vec::push(chars, 49) };
    set bitmap = i32_shr_u(bitmap, 1);
    set i = i + 1;
  };

  let char_count = chars.len;
  let bytes = bytes_new(0, char_count);
  let j = 0;
  while j < char_count {
    bytes_set(bytes, char_count - j - 1, chars.buf[j]);
    set j = j + 1;
  };
  string::unsafe_new(bytes)
}

fn print[a](hamt : Hamt[a]) -> string::Str {
  print_node#[a](hamt.root)
}

fn print_node[a](node : Node[a]) -> string::Str {
  match node {
    Node::Leaf l => {
      string::concat(string::concat(string::unsafe_new("(#leaf "), print_hash(l.hash)), string::unsafe_new(")"));
    },
    Node::Mapped m => {
      let nodes = string::unsafe_new("");
      let i = 0;
      while i < array_len#[Node[a]](m.nodes) {
        set nodes = string::concat(string::concat(nodes, string::unsafe_new(" ")), print_node(m.nodes[i]));
        set i = i + 1;
      };
      string::concat(
        string::concat(
          string::concat(
            string::unsafe_new("(#mapd "),
            print_bitmap(m.bitmap)),
          nodes),
        string::unsafe_new(")"));
    },
  };
}

struct Hamt[a] {
  root : Node::Mapped[a],
  size : I32,
}

variant Node[a] {
  struct Mapped { bitmap : I32, nodes : Array[Node[a]] },
  struct Leaf { hash : I32, value : a },
}

fn new[a]() -> Hamt[a] {
  Hamt { root = Node::Mapped { bitmap = 0, nodes = [] }, size = 0 }
}

fn get[a](hamt : Hamt[a], hash : I32) -> option::Option[a] {
  match get_with_anchor(hamt.root, 0, hash).payload {
    GetResultPayload::Success s => {
      option::Option::Some { val = s.node.value }
    },
    GetResultPayload::Missing _ => { option::Option::None {} },
    GetResultPayload::Conflict _ => { option::Option::None {} },
  };
}

fn add[a](hamt : Hamt[a], hash : I32, value : a) -> option::Option[a] {
  let get_result = get_with_anchor(hamt.root, 0, hash);
  match get_result.payload {
    GetResultPayload::Missing _ => {
      set hamt.size = hamt.size + 1;
      let n = get_result.anchor;
      let pos = bitpos(hash, get_result.shift);
      set n.bitmap = i32_or(n.bitmap, pos);
      let ix = index(n.bitmap, pos);
      set n.nodes = insert_node(n.nodes, ix, Node::Leaf { hash = hash, value = value });
      option::Option::None {};
    },
    GetResultPayload::Conflict c => {
      set hamt.size = hamt.size + 1;
      let n = get_result.anchor;
      let pos = bitpos(hash, get_result.shift);
      let ix = index(n.bitmap, pos);
      let new_node = merge_leafs(get_result.shift + 5, c.node, hash, value);
      set n.nodes[ix] = new_node;
      option::Option::None {};
    },
    GetResultPayload::Success s => {
      let prev = s.node.value;
      set s.node.value = value;
      option::Option::Some { val = prev };
    },
  };
}

fn remove[a](hamt : Hamt[a], hash : I32) -> option::Option[a] {
  match remove_rec(hamt.root, 0, hash) {
    RemoveResult::Success s => {
      set hamt.size = hamt.size - 1;
      option::Option::Some { val = s.node.value }
    },
    RemoveResult::Gathered g => {
      set hamt.size = hamt.size - 1;
      option::Option::Some { val = g.removed.value }
    },
    RemoveResult::NotFound n => { option::Option::None {} },
  };
}

fn find_any[a](node : Node[a]) -> option::Option[a] {
  match node {
    Node::Leaf l => { option::Option::Some { val = l.value } },
    Node::Mapped m => {
      let i = 0;
      let len = array_len#[Node[a]](m.nodes);
      while i < len {
        match find_any(m.nodes[i]) {
          option::Option::Some s => { return s },
          option::Option::None _ => { },
        };
        set i = i + 1;
      };
      option::Option::None {};
    },
  };
}

fn values[a](hamt : Hamt[a]) -> option::Option[vec::Vec[a]] {
  let zero = find_any#[a](hamt.root);
  let result = match zero {
    option::Option::Some s => { vec::new(s.val) },
    option::Option::None _ => { return option::Option::None {} },
  };
  values_node(result, hamt.root);
  option::Option::Some { val = result };
}

fn values_node[a](vec : vec::Vec[a], node : Node[a]) {
  match node {
    Node::Leaf l => { vec::push(vec, l.value) },
    Node::Mapped m => {
      let i = 0;
      let len = array_len#[Node[a]](m.nodes);
      while i < len {
        values_node(vec, m.nodes[i]);
        set i = i + 1;
      };
    },
  };
}

fn mask(hash : I32, shift : I32) -> I32 {
  i32_and(i32_shr_u(hash, shift), 31)
}

fn bitpos(hash : I32, shift : I32) -> I32 {
   i32_shl(1, mask(hash, shift))
}

fn index(bitmap : I32, pos : I32) -> I32 {
  i32_popcnt(i32_and(bitmap, (pos - 1)))
}

fn merge_leafs[a](shift : I32, left : Node::Leaf[a], hash : I32, value : a) -> Node[a] {
  let next_pos1 = bitpos(left.hash, shift);
  let next_pos2 = bitpos(hash, shift);
  if next_pos1 != next_pos2 {
    let bitmap = i32_or(next_pos1, next_pos2);
    let right = Node::Leaf { hash = hash, value = value };
    let nodes : Array[Node[a]] = if i32_lt_u(next_pos1, next_pos2) {
      [left, right]
    } else {
      [right, left]
    };
    Node::Mapped { bitmap = bitmap, nodes = nodes };
  } else {
    let new_node = merge_leafs(shift + 5, left, hash, value);
    Node::Mapped { bitmap = next_pos1, nodes = [new_node] };
  };
}

variant RemoveResult[a] {
  struct Success {
    node : Node::Leaf[a],
  },
  struct NotFound {},
  struct Gathered {
    new_node : Node::Leaf[a],
    removed : Node::Leaf[a],
  },
}

fn remove_rec[a](node : Node::Mapped[a], shift : I32, hash : I32) -> RemoveResult[a] {
  let pos = bitpos(hash, shift);
  if i32_and(pos, node.bitmap) == 0 {
    return RemoveResult::NotFound {};
  } else {};
  let ix = index(node.bitmap, pos);
  let next_node = node.nodes[ix];
  match next_node {
    Node::Mapped n => {
      let res = remove_rec(n, shift + 5, hash);
      match res {
        RemoveResult::Gathered gathered => {
          if shift != 0 && i32_popcnt(n.bitmap) == 1 {
            res;
          } else {
            set n.nodes[ix] = gathered.new_node;
            RemoveResult::Success { node = gathered.removed };
          };
        },
        _ => { res },
      };
    },
    Node::Leaf l => {
      if hash != l.hash {
        RemoveResult::NotFound {};
      } else {
        let rows = i32_popcnt(node.bitmap);
        // We never gather the root node
        if rows == 1 && shift == 0 {
          set node.bitmap = 0;
          set node.nodes = [];
          return RemoveResult::Success { node = l };
        } else {};
        if rows == 2 {
          let other = if ix == 0 { node.nodes[1] } else { node.nodes[0] };
          match other {
            Node::Leaf other_l => {
              if shift != 0 {
                return RemoveResult::Gathered { new_node = other_l, removed = l };
              } else {};
            },
            _ => {},
          };
          set node.bitmap = i32_and(node.bitmap, i32_xor(pos, 0 - 1));
          set node.nodes = [other];
          RemoveResult::Success { node = l };
        } else {
          let new_nodes = remove_node(node.nodes, ix);
          set node.bitmap = i32_and(node.bitmap, i32_xor(pos, 0 - 1));
          set node.nodes = new_nodes;
          RemoveResult::Success { node = l };
        };
      };
    },
  };
}

fn remove_node[a](as : Array[a], ix : I32) -> Array[a] {
  let old_len = array_len#[a](as);
  if old_len == 1 {
    [];
  } else {
    let res = array_new#[a](as[0], old_len - 1);
    array_copy#[a](res, 0, as, 0, ix);
    array_copy#[a](res, ix, as, ix + 1, old_len - ix - 1);
    res;
  };
}

fn insert_node[a](as : Array[a], ix : I32, elem : a) -> Array[a] {
  let old_len = array_len#[a](as);
  if old_len == 0 {
    [elem];
  } else {
    let res = array_new#[a](elem, old_len + 1);
    array_copy#[a](res, 0, as, 0, ix);
    array_copy#[a](res, ix + 1, as, ix, old_len - ix);
    res;
  };
}

struct GetResult[a] {
  shift : I32,
  anchor : Node::Mapped[a],
  payload : GetResultPayload[a],
}

variant GetResultPayload[a] {
  struct Success { node : Node::Leaf[a] },
  struct Conflict { node : Node::Leaf[a] },
  struct Missing {},
}

fn get_with_anchor[a](
  anchor : Node::Mapped[a],
  shift : I32,
  hash : I32,
) -> GetResult[a] {
  let pos = bitpos(hash, shift);
  if i32_and(anchor.bitmap, pos) == 0 {
    GetResult {
      shift = shift,
      anchor = anchor,
      payload = GetResultPayload::Missing {},
    };
  } else {
    let ix = index(anchor.bitmap, pos);
    let n = anchor.nodes[ix];
    match n {
      Node::Mapped anchor => {
        get_with_anchor(anchor, shift + 5, hash);
      },
      Node::Leaf leaf => {
        if leaf.hash == hash {
          GetResult {
            shift = shift,
            anchor = anchor,
            payload = GetResultPayload::Success { node = leaf },
          };
        } else {
          GetResult {
            shift = shift,
            anchor = anchor,
            payload = GetResultPayload::Conflict { node = leaf },
          };
        };
      },
    };
  };
}
