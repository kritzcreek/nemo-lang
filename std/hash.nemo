module hash
exports (FnvHasher, newHasher, write, write_i32, hash_i32, hash_bytes)

global fnv_prime = 16777619
// 2166136261 as a signed integer
global fnv_offset_basis = 0 - 2128831035

// http://www.isthe.com/chongo/tech/comp/fnv/#FNV-1a
fn fnv_1a(hash : I32, input : Bytes) -> I32 {
  let len = bytes_len(input);
  let i = 0;
  while i < len {
    set hash = i32_xor(hash, bytes_get(input, i)) * fnv_prime;
    set i = i + 1;
  };
  hash
}

fn fnv_1a_i32(hash : I32, input : I32) -> I32 {
  set hash = i32_xor(hash, i32_and(input, 0xFF)) * fnv_prime;
  set hash = i32_xor(hash, i32_and(i32_shr_u(input, 8), 0xFF)) * fnv_prime;
  set hash = i32_xor(hash, i32_and(i32_shr_u(input, 16), 0xFF)) * fnv_prime;
  i32_xor(hash, i32_shr_u(input, 24)) * fnv_prime
}

fn hash_i32(i : I32) -> I32 {
  fnv_1a_i32(fnv_offset_basis, i)
}

fn hash_bytes(i : Bytes) -> I32 {
  fnv_1a(fnv_offset_basis, i)
}

struct FnvHasher { hash : I32 }

fn newHasher() -> FnvHasher {
  FnvHasher { hash = fnv_offset_basis }
}

fn write(hasher : FnvHasher, bytes : Bytes) {
  set hasher.hash = fnv_1a(hasher.hash, bytes)
}

fn write_i32(hasher : FnvHasher, int : I32) {
  set hasher.hash = fnv_1a_i32(hasher.hash, int)
}
