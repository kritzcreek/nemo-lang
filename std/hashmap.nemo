module hashmap
exports (HashMap, new, add, get)

use io
use hamt
use option

variant Bucket[k, v] {
  struct Single { key : k, value : v },
  struct Conflict { key : k, value : v, next : Bucket[k, v] },
}

fn bucket_key[k, v](b : Bucket[k, v]) -> k {
  match b {
    Bucket::Single s => { s.key },
    Bucket::Conflict c => { c.key },
  };
}

fn bucket_value[k, v](b : Bucket[k, v]) -> v {
  match b {
    Bucket::Single s => { s.value },
    Bucket::Conflict c => { c.value },
  };
}

fn bucket_set_value[k, v](b : Bucket[k, v], value : v) {
  match b {
    Bucket::Single s => { set s.value = value },
    Bucket::Conflict c => { set c.value = value },
  };
}

fn bucket_add[k, v](b : Bucket[k, v], key : k, value : v) -> Bucket[k, v] {
  Bucket::Conflict { key = key, value = value, next = b }
}

fn bucket_find[k, v](
  b : Bucket[k, v],
  key : k,
  eq : fn (k, k) -> Bool
) -> option::Option[Bucket[k, v]] {
  match b {
    Bucket::Single single => {
      if eq(single.key, key) {
        option::Option::Some { val = single };
      } else {
        option::Option::None {};
      }
    },
    Bucket::Conflict conflict => {
      if eq(conflict.key, key) {
        option::Option::Some { val = conflict };
      } else {
        bucket_find(conflict.next, key, eq)
      }
    },
  }
}

fn bucket_get[k, v](
  b : Bucket[k, v],
  key : k,
  eq : fn (k, k) -> Bool
) -> option::Option[v] {
  option::map(bucket_find(b, key, eq), \(b) { bucket_value(b) })
}

struct HashMap[k, v] {
  hash : fn (k) -> I32,
  eq : fn (k, k) -> Bool,
  hamt : hamt::Hamt[Bucket[k, v]],
  size : I32,
}

fn new[k, v](hash : fn (k) -> I32, eq : fn (k, k) -> Bool) -> HashMap[k, v] {
  HashMap { hash = hash, eq = eq, hamt = hamt::new(), size = 0 }
}

fn add[k, v](map : HashMap[k, v], key : k, value : v) -> option::Option[v] {
  let hash = map.hash(key);
  match hamt::get(map.hamt, hash) {
    option::Option::Some s => {
      match bucket_find(s.val, key, map.eq) {
        option::Option::Some conflict => {
          let old_value = bucket_value(conflict.val);
          bucket_set_value(conflict.val, value);
          option::Option::Some { val = old_value }
        },
        option::Option::None _ => {
          hamt::add(map.hamt, hash, bucket_add(s.val, key, value));
          set map.size = map.size + 1;
          option::Option::None {};
        }
      };
    },
    option::Option::None _ => {
      hamt::add(map.hamt, hash, Bucket::Single { key = key, value = value });
      set map.size = map.size + 1;
      option::Option::None {}
    },
  };
}

fn get[k, v](map : HashMap[k, v], key : k) -> option::Option[v] {
  let hash = map.hash(key);
  match hamt::get(map.hamt, hash) {
    option::Option::Some bucket => {
      bucket_get(bucket.val, key, map.eq)
    },
    option::Option::None _ => {
      option::Option::None {};
    },
  };
}
