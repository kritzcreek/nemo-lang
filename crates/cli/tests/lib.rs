use backend::compile_program;
use camino::{Utf8Path, Utf8PathBuf};
use insta::{assert_snapshot, glob};
use insta_cmd::{assert_cmd_snapshot, get_cargo_bin};
use std::fmt;
use std::process::Command;
use std::{fs, path::Path};
use wasmtime::{Caller, Config, Engine, Linker, Module, Store};

fn read_source_files(
    input_files: Vec<Utf8PathBuf>,
) -> Result<Vec<(Utf8PathBuf, String)>, std::io::Error> {
    input_files
        .into_iter()
        .map(|input_file| {
            let source = fs::read_to_string(&input_file)?;
            Ok((input_file, source))
        })
        .collect()
}

fn render_slash_path(path: &Utf8Path) -> String {
    assert!(path.is_relative());
    path.components()
        .map(|c| c.as_str())
        .collect::<Vec<_>>()
        .join("/")
}

fn check_args(paths: &[Utf8PathBuf]) -> Vec<String> {
    let mut args = vec!["check".to_string()];
    args.extend(paths.iter().map(|p| render_slash_path(p)));
    args
}

fn cli() -> Command {
    Command::new(get_cargo_bin("nemo"))
}

fn collect_test_input(path: &Path) -> (String, Vec<Utf8PathBuf>) {
    let path = Utf8Path::from_path(path).expect("Non-utf8 path: {path:?}");
    let out_name = path.file_stem().unwrap();
    let mut paths = vec![];
    if path.is_dir() {
        for entry in fs::read_dir(path).expect("Failed to list directory {path:?}") {
            let entry = entry.expect("Failed to read entry");
            let path = entry.path();
            paths.push(Utf8PathBuf::from_path_buf(path).expect("non-utf8 path: {path:?}"));
        }
    } else {
        paths.push(path.to_owned());
    };
    (out_name.to_string(), paths)
}

enum Run {
    Success { result: i32, output: Vec<String> },
    Trap { reason: String, output: Vec<String> },
}

impl fmt::Display for Run {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Run::Success { result, output } => {
                writeln!(f, "Success: {result}")?;
                for out in output {
                    writeln!(f, "{out}")?;
                }
            }
            Run::Trap { reason, output } => {
                writeln!(f, "Trapped: {reason}")?;
                for out in output {
                    writeln!(f, "{out}")?;
                }
            }
        };
        Ok(())
    }
}

fn wasmtime_runner(wasm: Vec<u8>) -> Run {
    let engine = Engine::new(Config::new().wasm_function_references(true).wasm_gc(true)).unwrap();
    let mut store = Store::new(&engine, vec![]);
    let mut linker = Linker::new(&engine);
    let module = Module::new(&engine, &wasm).unwrap();
    linker
        .func_wrap(
            "env",
            "log",
            |mut caller: Caller<'_, Vec<String>>, param: i32| {
                caller.data_mut().push(format!("{param}"));
                0
            },
        )
        .unwrap();
    linker
        .func_wrap(
            "env",
            "log_int",
            |mut caller: Caller<'_, Vec<String>>, param: i32| {
                caller.data_mut().push(format!("{param}"));
                0
            },
        )
        .unwrap();
    linker
        .func_wrap(
            "env",
            "log_float",
            |mut caller: Caller<'_, Vec<String>>, param: f32| {
                caller.data_mut().push(format!("{param}"));
                0
            },
        )
        .unwrap();
    linker
        .func_wrap(
            "env",
            "log_f32",
            |mut caller: Caller<'_, Vec<String>>, param: f32| {
                caller.data_mut().push(format!("{param}"));
                0
            },
        )
        .unwrap();
    linker
        .func_wrap(
            "env",
            "print_char",
            |mut caller: Caller<'_, Vec<String>>, param: u32| {
                let mut last_line = caller.data_mut().pop().unwrap_or_default();
                last_line.push_str(&format!("{}", char::from_u32(param).unwrap()));
                caller.data_mut().push(last_line);
                0
            },
        )
        .unwrap();
    linker
        .func_wrap("env", "random", |_: i32| -> f32 { 0.0 })
        .unwrap();

    let instance = linker.instantiate(&mut store, &module).unwrap();
    let main_export = instance
        .exports(&mut store)
        .find(|export| export.name().ends_with("main"))
        .unwrap();
    let main_func = main_export
        .into_func()
        .unwrap()
        .typed::<(), i32>(&mut store)
        .unwrap();
    match main_func.call(&mut store, ()) {
        Ok(result) => Run::Success {
            result,
            output: store.into_data(),
        },
        Err(err) => Run::Trap {
            reason: err.to_string(),
            output: store.into_data(),
        },
    }
}

fn test_run(path: &Path) {
    let (_, paths) = collect_test_input(path);
    let sources = read_source_files(paths).unwrap();
    let wasm = compile_program(&sources).unwrap();
    let output = wasmtime_runner(wasm);
    assert_snapshot!(output)
}

fn test_check(path: &Path) {
    let (_, paths) = collect_test_input(path);
    assert_cmd_snapshot!(cli().env("RUST_BACKTRACE", "0").args(check_args(&paths)));
}

#[test]
fn check() {
    let cwd = std::env::current_dir().unwrap();
    glob!("check/*", |path| {
        let path = path.strip_prefix(&cwd).unwrap();
        test_check(path);
    });
}

#[test]
fn run() {
    let cwd = std::env::current_dir().unwrap();
    glob!("run/*", |path| {
        let path = path.strip_prefix(&cwd).unwrap();
        test_run(path);
    });
}
