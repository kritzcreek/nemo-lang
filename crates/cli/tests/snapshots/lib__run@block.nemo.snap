---
source: crates/cli/tests/lib.rs
expression: output
input_file: crates/cli/tests/run/block.nemo
---
Success: 0
1
2
4
